### DISCUSSION TITLE
Multithreading within the auto builder?

### DISCUSSION THREAD

**Discussion by ghost** [eclipse-platform/eclipse.platform#284] [Q&A]
Currently the auto-builder is able to work on multiple (independent) projects in parallel, but each project itself is still compiled sequentially, regardless how big the project is.

I've come accross some rather old bug reports regarding this issue, like [Bug 126121](https://bugs.eclipse.org/bugs/show_bug.cgi?id=126121) or [Bug 227025](https://bugs.eclipse.org/bugs/show_bug.cgi?id=227025). Have any changes been made since then?

**Context**: We use the incremental builder in our product to generate byte code from custom entities. Those entities are stored in separate projects, with each project containing several hundred files at a time. With a purely sequential build, this takes quite a lot of time.

In order to increase the performance, our builder spawns multiple "worker" jobs, which perform the actual compilation. The auto-builder remains inactive during that time and only continues after those jobs have finished. This itself works for the most part.

**The issue**: Because the worker jobs run in separate threads, they raise the "interrupted" flag of the auto-builder, whenever the compilation result is stored. This makes it difficult to cancel the job, as the auto-builder schedules itself for execution again, whenever this flag is set. This then triggers the "worker" jobs, they then raise the "interrupt" flag, the user cancels the new job and the cycle repeats itself over and over again...

Reason being that the auto-builder reserves the entire workspace. meaning that any modifications to it are causing a conflict. Which is hardcoded and not something I can influence in a meaningful way. Even if those "worker" jobs are technically a part of it, it is not something that can be expressed by the current Job API.

As a note, this rescheduling is only done whenever an OperationCanceledException is thrown by the builder. Meaning a "workaround" is to simply convert this into a plain CoreException. However, I wonder if there is cleaner way to do it or whether I've simply missed a feature for exactly this use case.

**Comment by jukzi** [comment]
You could use multiple threads to compile and gather the results into the builder thread. The java builder works like that: only one thread writes into workspace, but multiple threads read.

**Reply by ghost** [reply]
An interesting idea. I can probably do some buffering to then perform the read and write operations in parallel. 
I'm not sure how difficult it will be to split the read and write operations in our implementation. But given that our class is in dire need of refactoring anyway, that might be a good thing. 😅
I'll have a look at the JavaBuilder and see if there are any ideas I can adapt.

Thanks.

**Comment by Bananeweizen** [comment]
How exactly are you writing your results to the workspace? I would expect that you can get an ISchedulingRule for the complete workspace, and write your results from within that scheduling rule, thereby effectively blocking further jobs running on the same projects. But I haven't actually tested this, I just remember we had the opposite of your issue (not enough stuff being triggered in parallel because of too large scheduling rule).

**Reply by ghost** [reply]
We use IFile.create() to store the data. As its scheduling rule, we use the parent directory (e.g. "/Production/bin/buildpackage").
But even if I set it to the workspace root, the same issue shows up.

From what I can tell, the IFile implementation always checks its scheduling rule (its relative path) against the auto-builder via IWorkspace.pepareOperation(). Whenever the auto-builder is running, the interrupt flag is then raised. This seems to be a hardcoded behaviour which can't be influenced from the outside.
And given that our worker threads are running together with the auto-builder, this means the flag is then always raised.

**Comment by ghost** [comment]
Thank you both for your responses.

After digging through both our and the auto-build implementation, I don't think it's possible to prevent the auto-build job from rescheduling itself over and over again.

Given the size of our builders, I also don't see an easy way for splitting the read and write without a major time investment and without re-creating this problem, whenever we do changes there.
So ideally, the Eclipse API should give the user a little bit more control over the auto-builder. 

My question: Would there be an interest in me creating a pull request to address those issues?

Possible solutions I came up with are:

(I) Introduce a new property which can be assigned to jobs. The interrupt flag is currently set on all workspace modifications, unless those were caused auto-build job itself. External jobs could then opt-in (at their own risk, of course) by setting this property, so that their modifications are also ignored.

(II) Introduce a new job state. Our worker jobs are inside a job group. From within the AutoBuild job, we then call join(), to wait for those worker jobs to complete. While waiting, the auto-build job would then switch to an "idle" state, during which interrupts are ignored.

I'm leaning towards the first option, as we can reuse the setProperty() method of the Job API and because it shouldn't cause any unintended side-effects for code that doesn't use this functionality. Thoughts?

**Comment by iloveeclipse** [comment]
> Because the worker jobs run in separate threads, they raise the "interrupted" flag of the auto-builder

Can you please provide a stack showing how exactly this happens?
In general, a simple example project that reproduces your problem would help to understand it and give any meaningful advice.

> Would there be an interest in me creating a pull request to address those issues?

If the problem is clearly stated & understood. I'm not sure this is the case.

My understanding is that you want to generate something in parallel during the autobuild and write it via resources API **synchronously** to the build invocation that locked **same** project. I wonder if you can postpone the write after the build is done or use Java NIO API or write **from builder thread** (put lambdas that write on the queue of your builder)?

**Reply by ghost** [reply]
> you want to generate something in parallel during the autobuild and write it via resources API synchronously to the build invocation that locked same project

That is true. The auto-build is inactive during that time and the remaining synchronization between those threads is handeled by us.

>  I wonder if you can postpone the write after the build [...] or write from builder thread

I thought about that as well, but I don't think that's really possible in our case. The parallelization is done by a delegator, which then calls the correct builder for each file. The builders themself don't know, whether they are in the auto-build job or in a separate thread. 

> or use Java NIO API

That definitely works. Though it feels wrong to _not_ use the Eclipse API when modifying on the workspace.
We're also setting a resource property for some of the files. I haven't tested whether that would also cause an interrupt.

Here is a simple builder implementation to illustrate the issue. You can add/remove this builder via the Properties -> Configure tab.  It's currently configured to cancel the auto-build upon completion (to simulate a user cancellation), leading to an infinite loop.

[custombuilder.zip](https://github.com/eclipse-platform/eclipse.platform/files/10132243/custombuilder.zip)

![image](https://user-images.githubusercontent.com/92990747/205054175-3a2f3e22-c44e-4c05-a880-b8c7d0f3754d.png)

Edit:

In addition, the following method in the AutoBuilderJob class is responsible for this behaviour:
```
private synchronized IStatus canceled() {
	//regardless of the form of cancelation, the build state is not happy
	buildNeeded = true;
	//schedule a rebuild immediately if build was implicitly canceled
	if (interrupted) {
		if (Policy.DEBUG_BUILD_INTERRUPT)
			Policy.debug("Scheduling rebuild due to interruption"); //$NON-NLS-1$
		setInterrupted(false);
		schedule(computeScheduleDelay());
	}
	return Status.CANCEL_STATUS;
}
```

The interrupt flag is set by Workspace.prepareOperation(), which is called by the worker threads.

**Comment by jukzi** [comment]
I don't see any reason to extend eclipse for parallel builder support, as (at least on windows) the main performance bootleneck is always writing the result, and currently the resource API just can't write concurrently. You are better off fixing your builder.
If you found an error in the autobuild cancel please start with writing a junit test that shows it and then provide a fix.

**Reply by ghost** [reply]
> the main performance bootleneck is always writing the result, and currently the resource API just can't write concurrently.

That is simply not true. With two threads, we have a performance improvement of roughly 40% and with four threads, it's somewhere around 60%. So going sequential isn't really an alternative.

And I'm not sure the problem really lies in our builder implementation. In the end, it works perfectly fine on both a full and incremental build, and even with the automatic build, **unless** you want to cancel the job. Then the build starts over and over again.
I've created #289  to illustrate the problem.

But as a workaround, I will most likely just disable the auto-builder, whenever the build job has been cancelled.
